src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -940,48 +1005,53 @@ result::Install SotaUptaneClient::uptaneInstall(const std::vector<Uptane::Target
-        report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(ecu_serial, correlation_id, true));
-        sendEvent<event::InstallTargetComplete>(ecu_serial, true);
+        // note: installation result is not save here, will be done after finalization
+        report_queue->enqueue(std_::make_unique<EcuInstallationAppliedReport>(primary_ecu_serial, correlation_id));
+        sendEvent<event::InstallTargetComplete>(primary_ecu_serial, true);  // TODO: distinguish from success here?
+      } else if (install_res.result_code.num_code == data::ResultCode::Numeric::kOk) {
+        storage->saveEcuInstallationResult(primary_ecu_serial, install_res);
+        report_queue->enqueue(
+            std_::make_unique<EcuInstallationCompletedReport>(primary_ecu_serial, correlation_id, true));
+        sendEvent<event::InstallTargetComplete>(primary_ecu_serial, true);
       } else {
         // general error case
-        report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(ecu_serial, correlation_id, false));
-        sendEvent<event::InstallTargetComplete>(ecu_serial, false);
+        storage->saveEcuInstallationResult(primary_ecu_serial, install_res);
+        report_queue->enqueue(
+            std_::make_unique<EcuInstallationCompletedReport>(primary_ecu_serial, correlation_id, false));
+        sendEvent<event::InstallTargetComplete>(primary_ecu_serial, false);
       }
     } else {
-      data::InstallOutcome outcome(data::UpdateResultCode::kAlreadyProcessed, "Package already installed");
-      status = data::OperationResult(primary_update.filename(), outcome);
-      storage->storeInstallationResult(status);
+      install_res = data::InstallationResult(data::ResultCode::Numeric::kAlreadyProcessed, "Package already installed");
+      storage->saveEcuInstallationResult(primary_ecu_serial, install_res);
       // TODO: distinguish this case from regular failure for local and remote
       // event reporting
       report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(uptane_manifest.getPrimaryEcuSerial(),
                                                                               correlation_id, false));
       sendEvent<event::InstallTargetComplete>(uptane_manifest.getPrimaryEcuSerial(), false);
     }
-    result.reports.emplace(result.reports.begin(), primary_update, uptane_manifest.getPrimaryEcuSerial(), status);
+    result.ecu_reports.emplace(result.ecu_reports.begin(), primary_update, uptane_manifest.getPrimaryEcuSerial(),
+                               install_res);
     // TODO: other updates for primary
   } else {
     LOG_INFO << "No update to install on primary";
   }
 
   auto sec_reports = sendImagesToEcus(updates);
-  result.reports.insert(result.reports.end(), sec_reports.begin(), sec_reports.end());
+  result.ecu_reports.insert(result.ecu_reports.end(), sec_reports.begin(), sec_reports.end());
+  computeDeviceInstallationResult(&result.dev_report, correlation_id);
```

Comment:
So it's a bit tricky. At install(), a package manager can return:

- kOk. If all ecus are ok, device installation result will be ok and the manifest will be sent immediately (since there is no pending update)
- kNeedCompletion. If one ecu needs completion, the manifest won't be sent until the flag is unset which will happen at next reboot. Then device result will be ok or error depending on the final states of all ecus
- an error. If one ecu has an error, device installation result will not be ok. But one ecu needing completion will still block manifest emission.

Don't know if it's really clear, maybe it would need a state machine drawing. Also, kNeedCompletion will only be supported for primaries right now.